Feature/architecture

AElkhami opened the pull request.

Commit:
*Creating Data layer.
*Moving UI classes to presentation package.
*Creating di.

Commit:
*Creating domain layer.
*Creating domain module for di.
*Adding mockito dependency.
*Moving base url to build config.

Commit:
*Creating presentation layer.
*Creating home screen UI.
*Adding coil dependency.
*Adding internet permission.

Commit:
*Updating app dimensions and spacing.

Commit:
*Implementing viewModel with coverage.

Commit:
*ViewModel connect with UI.
*Minor fixes.
*App is now working.

Commit:
*Implement loading on UI.
*Minor fixes.

Commit:
*Moving data logic from use case to repository.

Commit:
*Moving Auth_Key to build config.

Commit:
*Migrating from mockito to mockk.

Commit:
*Adding repository unit test.

Commit:
*Adding exception handling to Api calls.
*Moving data change logic from repository to viewModel.
*Updating unit tests.
*Updating ktor library.
*Removing unnecessary class MovieService.kt

Commit:
*Adding toggle button.
*Adding horizontal movie view.
*Updating home screen.
*Minor refactoring.

Commit:
*Checking for internet connection.
*Creating error view.
*Bug fixes and enhancements.

Commit:
*Moving some values to constants.

MohamedHashim commented on app/build.gradle:
```diff
@@ -21,7 +22,16 @@ android {
 
     buildTypes {
+        debug {
+            buildConfigField "String", "AUTH_KEY","\"Bearer eyJhbGciOiJIUzI1NiJ9.eyJhdWQiOiJjZWVkZTk2ZDUxNTQ4NTA1ZTY3YTY4NjFhY2U3NjBmYyIsInN1YiI6IjY0YjY1MmUxMzc4MDYyMDBmZjNhNWM0NyIsInNjb3BlcyI6WyJhcGlfcmVhZCJdLCJ2ZXJzaW9uIjoxfQ.g3GD-e6kvUmXyPXG8EJ8lAYPBEDZQy4LQzrmn8rngBA\""
```
From a security perspective, We shouldn't reveal the API key like this as it's exposed to the public and anyone can use it.
Try to add this key in gradle.properties file and then add this file to .gitignore file to make it invisible in the version control.

MohamedHashim commented on app/src/main/java/com/example/aimovies/presentation/home/HomeScreen.kt:
```diff
@@ -0,0 +1,178 @@
+ */
+
+val movies = listOf(
+    MovieModel(
+        overview = "test",
+        title = "The Tomorrow War",
+        voteCount = 10,
+        voteAverage = 8.5,
+        posterPath = "https://assets-global.website-files.com/6009ec8cda7f305645c9d91b/6408f676b5811234c887ca62_top%20gun%-min.png",
+        releaseDate = "01/03/2023"
+    ), MovieModel(
+        overview = "test",
+        title = "The Tomorrow War",
+        voteCount = 10,
+        voteAverage = 8.5,
+        posterPath = "https://assets-global.website-files.com/6009ec8cda7f305645c9d91b/6408f676b5811234c887ca62_top%20gun%20maverick-min.png",
+        releaseDate = "01/03/2023"
+    ), MovieModel(
+        overview = "test",
+        title = "The Tomorrow War",
+        voteCount = 10,
+        voteAverage = 8.5,
+        posterPath = "https://assets-global.website-files.com/6009ec8cda7f305645c9d91b/6408f676b5811234c887ca62_top%20gun%20maverick-min.png",
+        releaseDate = "01/03/2023"
+    )
+)
```
Why do we need to have static movies while getting them through API?

Commit:
*Creating sqlDelight db for saving favourites.
*Implementing the required crud functions.
*Adding sqlDelight dependencies to gradle.
*Creating overview screen.

Commit:
*Package refactoring.
*Adding compose navigation dependency.
*Implementing compose navigation.
*Implementing overview screen and viewModel.

Commit:
*Implementing remove favourite in overview screen.
*Displaying different color based weather the movie is favourite or not.

Commit:
*Minor fixes.

Commit:
*Moving auth key to local properties.

Commit:
*Creating new tests to cover created classes.
*Increasing test coverage.

Commit:
*Finalizing unit tests and increasing test coverage.
*Minor refactoring.
*Adding sqlDelight testing library into gradle.

Commit:
*UI enhancements.

MohamedHashim commented on app/src/main/java/com/example/aimovies/presentation/home/HomeScreen.kt:
```diff
@@ -171,7 +146,7 @@ fun HomeScreenUi(
 fun HomeScreenPreview() {
     AIMoviesTheme {
-        HomeScreenUi(HomeUiModel(movies)) {
+        HomeScreenUi(HomeUiModel(emptyList())) {
```
You can use HomeUiModel() without emptyList() as it has a default value of emptyList().

MohamedHashim commented on app/src/main/java/com/example/aimovies/domain/use_case/GetFavouriteMovies.kt:
```diff
@@ -0,0 +1,14 @@
+ * Created by A.Elkhami on 25/07/2023.
+ */
+class GetFavouriteMovies(private val repository: FavouriteMovieRepository) {
+    operator fun invoke(): Flow<List<FavouriteEntitiy>> {
```
I think this function should be suspend function.

AElkhami replied:
This function returns flow and doesn't call any other suspend function, so it doesn't need to be suspended as the flow will do anything related async within itself.

MohamedHashim reviewed (approved):
Good job!

Commit:
*Minor enhancement.
